Unable to connect to the ES when using K8s secrets

HI,

I have setup the ES cleaner cronjob for an externally hosted Elastic search cluster, and i am trying to use secrets stored in kubernetes to connect to the ES cluster.

Below is my yaml
```apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: es-jaeger-cleaner
  namespace: monitoring
  labels:
    app: es-jaeger-cleaner
    env: staging
spec:
  # every 1 PM UTC-0
  schedule: "0 13 * * *"
  jobTemplate:
    metadata:
      labels:
        app: es-jaeger-cleaner
        env: staging
    spec:
      template:
        metadata:
          labels:
            app: es-jaeger-cleaner
            env: staging
        spec:
          containers:
          - name: es-jaeger-cleaner
            image: jaegertracing/jaeger-es-index-cleaner:latest
            # clean up ES data indices older than 7 days from now
            args: ["7", "<https://<esHOST:port>"]
            env:
            - name: TIMEOUT
              value: "300"
            - name: ES_USERNAME
              value: "<username>"
            - name: ES_PASSWORD
              valueFrom:
                secretKeyRef:
                  key: ES_PASSWORD
                  name: <secret_name>
          restartPolicy: OnFailure
```

And it somehow does not work and i get the below error

```
File "/es-index-cleaner/esCleaner.py", line 31, in main
    client = elasticsearch.Elasticsearch(sys.argv[2:], http_auth=(username, password))
  File "/usr/local/lib/python3.7/site-packages/elasticsearch/client/__init__.py", line 206, in __init__
    self.transport = transport_class(_normalize_hosts(hosts), **kwargs)
  File "/usr/local/lib/python3.7/site-packages/elasticsearch/transport.py", line 127, in __init__
    self.set_connections(hosts)
  File "/usr/local/lib/python3.7/site-packages/elasticsearch/transport.py", line 177, in set_connections
    connections = list(zip(connections, hosts))
  File "/usr/local/lib/python3.7/site-packages/elasticsearch/transport.py", line 173, in _create_connection
    return self.connection_class(**kwargs)
  File "/usr/local/lib/python3.7/site-packages/elasticsearch/connection/http_urllib3.py", line 106, in __init__
    self.headers.update(urllib3.make_headers(basic_auth=http_auth))
  File "/usr/local/lib/python3.7/site-packages/urllib3/util/request.py", line 65, in make_headers
    b64encode(b(basic_auth)).decode('utf-8')
  File "/usr/local/lib/python3.7/site-packages/urllib3/packages/six.py", line 620, in b
    return s.encode("latin-1")
UnicodeEncodeError: 'latin-1' codec can't encode characters in position 9-13: ordinal not in range(256)
```

PS - If i hardcode the username and password, it works fine.
Please let me know if i am doing something wrong

Hello @pavolloffay. Thank you very much for your quick reaction!

Yes, I've seen many people having this issue when using the operator and, after reading some posts and threads, I haven't found any workaround or final solution, that´s why I'm calling for help :). 

I have tried pretty much every configuration and some of the changes proposed in the operator to make sure it mounts volumes and secrets required on the cronjob containers. Jaeger Ingester, Collector and Query connect to ES with no issue. Based on my understanding, the configuration should be the same as for the cronjobs (Operator gets it from storage options). I see username and password should be used and I'm passing these from a secret to the jobs via operator, together with the certs I've got from elasticsearch. However, the jobs seem to keep failing. Please find below the logs from the jobs, as per your request.

Regarding the point you've made about disabling sparkjob when TLS is enabled, don´t we need spark jobs running every day, same as for the index cleaner?

If we were to run spark and index cleaner cronjobs sepparately (flagging them as enable=false in the operator config), do you have any sample yaml I could use where TLS is used to connect to ES? 

Thank you very much! 

```
[root@devvm jaeger-operator]# kubectl logs pubsub-streaming-spark-dependencies-1573343700-2jvwd -n tracing
19/11/10 08:27:11 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
19/11/10 08:27:12 INFO ElasticsearchDependenciesJob: Running Dependencies job for 2019-11-10T00:00Z, reading from jaeger-span-2019-11-10 index, result storing to jaeger-dependencies-2019-11-10
19/11/10 08:27:13 ERROR NetworkClient: Node [https://10.102.180.152:9200] failed (javax.net.ssl.SSLHandshakeException: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target); no other nodes left - aborting...
Exception in thread "main" org.elasticsearch.hadoop.EsHadoopIllegalArgumentException: Cannot detect ES version - typically this happens if the network/Elasticsearch cluster is not accessible or when targeting a WAN/Cloud instance without the proper setting 'es.nodes.wan.only'
	at org.elasticsearch.hadoop.rest.InitializationUtils.discoverClusterInfo(InitializationUtils.java:340)
	at org.elasticsearch.hadoop.rest.RestService.findPartitions(RestService.java:220)
	at org.elasticsearch.spark.rdd.AbstractEsRDD.esPartitions$lzycompute(AbstractEsRDD.scala:79)
	at org.elasticsearch.spark.rdd.AbstractEsRDD.esPartitions(AbstractEsRDD.scala:78)
	at org.elasticsearch.spark.rdd.AbstractEsRDD.getPartitions(AbstractEsRDD.scala:48)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:253)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:251)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:251)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:253)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:251)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:251)
	at org.apache.spark.Partitioner$$anonfun$4.apply(Partitioner.scala:75)
	at org.apache.spark.Partitioner$$anonfun$4.apply(Partitioner.scala:75)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.immutable.List.map(List.scala:285)
	at org.apache.spark.Partitioner$.defaultPartitioner(Partitioner.scala:75)
	at org.apache.spark.rdd.RDD$$anonfun$groupBy$1.apply(RDD.scala:691)
	at org.apache.spark.rdd.RDD$$anonfun$groupBy$1.apply(RDD.scala:691)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:363)
	at org.apache.spark.rdd.RDD.groupBy(RDD.scala:690)
	at org.apache.spark.api.java.JavaRDDLike$class.groupBy(JavaRDDLike.scala:243)
	at org.apache.spark.api.java.AbstractJavaRDDLike.groupBy(JavaRDDLike.scala:45)
	at io.jaegertracing.spark.dependencies.elastic.ElasticsearchDependenciesJob.run(ElasticsearchDependenciesJob.java:224)
	at io.jaegertracing.spark.dependencies.elastic.ElasticsearchDependenciesJob.run(ElasticsearchDependenciesJob.java:203)
	at io.jaegertracing.spark.dependencies.DependenciesSparkJob.run(DependenciesSparkJob.java:54)
	at io.jaegertracing.spark.dependencies.DependenciesSparkJob.main(DependenciesSparkJob.java:40)
Caused by: org.elasticsearch.hadoop.rest.EsHadoopNoNodesLeftException: Connection error (check network and/or proxy settings)- all nodes failed; tried [[https://10.102.180.152:9200]] 
	at org.elasticsearch.hadoop.rest.NetworkClient.execute(NetworkClient.java:152)
	at org.elasticsearch.hadoop.rest.RestClient.execute(RestClient.java:424)
	at org.elasticsearch.hadoop.rest.RestClient.execute(RestClient.java:388)
	at org.elasticsearch.hadoop.rest.RestClient.execute(RestClient.java:392)
	at org.elasticsearch.hadoop.rest.RestClient.get(RestClient.java:168)
	at org.elasticsearch.hadoop.rest.RestClient.mainInfo(RestClient.java:735)
	at org.elasticsearch.hadoop.rest.InitializationUtils.discoverClusterInfo(InitializationUtils.java:330)
	... 33 more
```
```
 [root@devvm jaeger-operator]# kubectl logs pubsub-streaming-es-index-cleaner-1573343700-lpngp -n tracing
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 345, in _make_request
    self._validate_conn(conn)
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 844, in _validate_conn
    conn.connect()
  File "/usr/local/lib/python3.7/site-packages/urllib3/connection.py", line 326, in connect
    ssl_context=context)
  File "/usr/local/lib/python3.7/site-packages/urllib3/util/ssl_.py", line 325, in ssl_wrap_socket
    return context.wrap_socket(sock, server_hostname=server_hostname)
  File "/usr/local/lib/python3.7/ssl.py", line 423, in wrap_socket
    session=session
  File "/usr/local/lib/python3.7/ssl.py", line 870, in _create
    self.do_handshake()
  File "/usr/local/lib/python3.7/ssl.py", line 1139, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1076)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/elasticsearch/connection/http_urllib3.py", line 233, in perform_request
    method, url, body, retries=Retry(False), headers=request_headers, **kw
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 630, in urlopen
    raise SSLError(e)
urllib3.exceptions.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1076)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/curator/utils.py", line 643, in get_indices
    index='_all', params={'expand_wildcards': 'open,closed'})
  File "/usr/local/lib/python3.7/site-packages/elasticsearch/client/utils.py", line 84, in _wrapped
    return func(*args, params=params, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/elasticsearch/client/indices.py", line 643, in get_settings
    "GET", _make_path(index, "_settings", name), params=params
  File "/usr/local/lib/python3.7/site-packages/elasticsearch/transport.py", line 350, in perform_request
    timeout=timeout,
  File "/usr/local/lib/python3.7/site-packages/elasticsearch/connection/http_urllib3.py", line 242, in perform_request
    raise SSLError("N/A", str(e), e)
elasticsearch.exceptions.SSLError: ConnectionError([SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1076)) caused by: SSLError([SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1076))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/es-index-cleaner/esCleaner.py", line 106, in <module>
    main()
  File "/es-index-cleaner/esCleaner.py", line 40, in main
    ilo = curator.IndexList(client)
  File "/usr/local/lib/python3.7/site-packages/curator/indexlist.py", line 31, in __init__
    self.__get_indices()
  File "/usr/local/lib/python3.7/site-packages/curator/indexlist.py", line 66, in __get_indices
    self.all_indices = utils.get_indices(self.client)
  File "/usr/local/lib/python3.7/site-packages/curator/utils.py", line 653, in get_indices
    raise exceptions.FailedExecution('Failed to get indices. Error: {0}'.format(e))
curator.exceptions.FailedExecution: Failed to get indices. Error: ConnectionError([SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1076)) caused by: SSLError([SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1076))

```

Thank you very much, @pavolloffay!

--- a/pkg/cronjob/es_index_cleaner.go
+++ b/pkg/cronjob/es_index_cleaner.go
@@ -24,17 +24,8 @@ func CreateEsIndexCleaner(jaeger *v1.Jaeger) *batchv1beta1.CronJob {
 	one := int32(1)
 	name := fmt.Sprintf("%s-es-index-cleaner", jaeger.Name)
 
-	var envFromSource []corev1.EnvFromSource
-	if len(jaeger.Spec.Storage.SecretName) > 0 {
-		envFromSource = append(envFromSource, corev1.EnvFromSource{
-			SecretRef: &corev1.SecretEnvSource{
-				LocalObjectReference: corev1.LocalObjectReference{
-					Name: jaeger.Spec.Storage.SecretName,
-				},
-			},
-		})
-	}
-	envs := esScriptEnvVars(jaeger.Spec.Storage.Options)
+	envFromSource := util.CreateEnvsFromSecret(jaeger.Spec.Storage.SecretName)
+	envs := EsScriptEnvVars(jaeger.Spec.Storage.Options)
 	if val, ok := jaeger.Spec.Storage.Options.Map()["es.use-aliases"]; ok && strings.EqualFold(val, "true") {
 		envs = append(envs, corev1.EnvVar{Name: "ROLLOVER", Value: "true"})
 	}
@@ -59,9 +50,10 @@ func CreateEsIndexCleaner(jaeger *v1.Jaeger) *batchv1beta1.CronJob {
 
 	return &batchv1beta1.CronJob{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      name,
-			Namespace: jaeger.Namespace,
-			Labels:    commonSpec.Labels,
+			Name:        name,
+			Namespace:   jaeger.Namespace,
+			Labels:      commonSpec.Labels,
+			Annotations: commonSpec.Annotations,
 			OwnerReferences: []metav1.OwnerReference{
 				{
 					APIVersion: jaeger.APIVersion,
@@ -83,12 +75,13 @@ func CreateEsIndexCleaner(jaeger *v1.Jaeger) *batchv1beta1.CronJob {
 						Spec: corev1.PodSpec{
 							Containers: []corev1.Container{
 								{
-									Name:      name,
-									Image:     jaeger.Spec.Storage.EsIndexCleaner.Image,
-									Args:      []string{strconv.Itoa(*jaeger.Spec.Storage.EsIndexCleaner.NumberOfDays), esUrls},
-									Env:       envs,
-									EnvFrom:   envFromSource,
-									Resources: commonSpec.Resources,
+									Name:         name,
+									Image:        jaeger.Spec.Storage.EsIndexCleaner.Image,
+									Args:         []string{strconv.Itoa(*jaeger.Spec.Storage.EsIndexCleaner.NumberOfDays), esUrls},
+									Env:          util.RemoveEmptyVars(envs),
+									EnvFrom:      envFromSource,
+									Resources:    commonSpec.Resources,
+									VolumeMounts: commonSpec.VolumeMounts,
 								},
 							},
 							RestartPolicy:      corev1.RestartPolicyNever,
@@ -96,6 +89,7 @@ func CreateEsIndexCleaner(jaeger *v1.Jaeger) *batchv1beta1.CronJob {
 							Tolerations:        commonSpec.Tolerations,
 							SecurityContext:    commonSpec.SecurityContext,
 							ServiceAccountName: account.JaegerServiceAccountFor(jaeger, account.EsIndexCleanerComponent),
+							Volumes:            commonSpec.Volumes,
 						},
 						ObjectMeta: metav1.ObjectMeta{
 							Labels:      commonSpec.Labels,

--- a/pkg/cronjob/es_rollover.go
+++ b/pkg/cronjob/es_rollover.go
@@ -31,7 +31,7 @@ func CreateRollover(jaeger *v1.Jaeger) []batchv1beta1.CronJob {
 
 func rollover(jaeger *v1.Jaeger) batchv1beta1.CronJob {
 	name := fmt.Sprintf("%s-es-rollover", jaeger.Name)
-	envs := esScriptEnvVars(jaeger.Spec.Storage.Options)
+	envs := EsScriptEnvVars(jaeger.Spec.Storage.Options)
 	if jaeger.Spec.Storage.EsRollover.Conditions != "" {
 		envs = append(envs, corev1.EnvVar{Name: "CONDITIONS", Value: jaeger.Spec.Storage.EsRollover.Conditions})
 	}
@@ -59,6 +59,7 @@ func rollover(jaeger *v1.Jaeger) batchv1beta1.CronJob {
 }
 
 func createTemplate(name, action string, jaeger *v1.Jaeger, envs []corev1.EnvVar) *corev1.PodTemplateSpec {
+	envFromSource := util.CreateEnvsFromSecret(jaeger.Spec.Storage.SecretName)
 	baseCommonSpec := v1.JaegerCommonSpec{
 		Annotations: map[string]string{
 			"prometheus.io/scrape":    "false",
@@ -80,13 +81,16 @@ func createTemplate(name, action string, jaeger *v1.Jaeger, envs []corev1.EnvVar
 			Tolerations:        commonSpec.Tolerations,
 			SecurityContext:    commonSpec.SecurityContext,
 			ServiceAccountName: account.JaegerServiceAccountFor(jaeger, account.EsRolloverComponent),
+			Volumes:            commonSpec.Volumes,
 			Containers: []corev1.Container{
 				{
-					Name:      name,
-					Image:     jaeger.Spec.Storage.EsRollover.Image,
-					Args:      []string{action, util.GetEsHostname(jaeger.Spec.Storage.Options.Map())},
-					Env:       envs,
-					Resources: commonSpec.Resources,
+					Name:         name,
+					Image:        jaeger.Spec.Storage.EsRollover.Image,
+					Args:         []string{action, util.GetEsHostname(jaeger.Spec.Storage.Options.Map())},
+					Env:          util.RemoveEmptyVars(envs),
+					EnvFrom:      envFromSource,
+					Resources:    commonSpec.Resources,
+					VolumeMounts: commonSpec.VolumeMounts,
 				},
 			},
 		},
@@ -95,7 +99,7 @@ func createTemplate(name, action string, jaeger *v1.Jaeger, envs []corev1.EnvVar
 
 func lookback(jaeger *v1.Jaeger) batchv1beta1.CronJob {
 	name := fmt.Sprintf("%s-es-lookback", jaeger.Name)
-	envs := esScriptEnvVars(jaeger.Spec.Storage.Options)
+	envs := EsScriptEnvVars(jaeger.Spec.Storage.Options)
 	if jaeger.Spec.Storage.EsRollover.ReadTTL != "" {
 		dur, err := time.ParseDuration(jaeger.Spec.Storage.EsRollover.ReadTTL)
 		if err == nil {
@@ -131,16 +135,27 @@ func lookback(jaeger *v1.Jaeger) batchv1beta1.CronJob {
 	}
 }
 
-func esScriptEnvVars(opts v1.Options) []corev1.EnvVar {
+// EsScriptEnvVars returns environmental variables for ES cron jobs.
+func EsScriptEnvVars(opts v1.Options) []corev1.EnvVar {
+	scriptEnvVars := []struct {
+		flag   string
+		envVar string
+	}{
+		{flag: "es.index-prefix", envVar: "INDEX_PREFIX"},
+		{flag: "es.username", envVar: "ES_USERNAME"},
+		{flag: "es.password", envVar: "ES_PASSWORD"},
+		{flag: "es.tls", envVar: "ES_TLS"},
+		{flag: "es.tls.ca", envVar: "ES_TLS_CA"},
+		{flag: "es.tls.cert", envVar: "ES_TLS_CERT"},
+		{flag: "es.tls.key", envVar: "ES_TLS_KEY"},
+		{flag: "es.tls.skip-host-verify", envVar: "ES_TLS_SKIP_HOST_VERIFY"},
+	}
+	options := opts.Map()
 	var envs []corev1.EnvVar
-	if val, ok := opts.Map()["es.index-prefix"]; ok {
-		envs = append(envs, corev1.EnvVar{Name: "INDEX_PREFIX", Value: val})
-	}
-	if val, ok := opts.Map()["es.username"]; ok {
-		envs = append(envs, corev1.EnvVar{Name: "ES_USERNAME", Value: val})
-	}
-	if val, ok := opts.Map()["es.password"]; ok {
-		envs = append(envs, corev1.EnvVar{Name: "ES_PASSWORD", Value: val})
+	for _, x := range scriptEnvVars {
+		if val, ok := options[x.flag]; ok {
+			envs = append(envs, corev1.EnvVar{Name: x.envVar, Value: val})
+		}
 	}
 	return envs
 }

--- a/pkg/cronjob/es_rollover_test.go
+++ b/pkg/cronjob/es_rollover_test.go
@@ -71,17 +71,29 @@ func TestEnvVars(t *testing.T) {
 		},
 		{
 			opts: v1.NewOptions(map[string]interface{}{
-				"es.index-prefix": "foo",
-				"es.password":     "nopass",
-				"es.username":     "fredy"}),
+				"es.index-prefix":         "foo",
+				"es.password":             "nopass",
+				"es.username":             "fredy",
+				"es.tls":                  "true",
+				"es.tls.ca":               "/etc/ca",
+				"es.tls.key":              "/etc/key",
+				"es.tls.cert":             "/etc/cert",
+				"es.tls.skip-host-verify": "true",
+			}),
 			expected: []corev1.EnvVar{
 				{Name: "INDEX_PREFIX", Value: "foo"},
 				{Name: "ES_USERNAME", Value: "fredy"},
-				{Name: "ES_PASSWORD", Value: "nopass"}},
+				{Name: "ES_PASSWORD", Value: "nopass"},
+				{Name: "ES_TLS", Value: "true"},
+				{Name: "ES_TLS_CA", Value: "/etc/ca"},
+				{Name: "ES_TLS_CERT", Value: "/etc/cert"},
+				{Name: "ES_TLS_KEY", Value: "/etc/key"},
+				{Name: "ES_TLS_SKIP_HOST_VERIFY", Value: "true"},
+			},
 		},
 	}
 	for _, test := range tests {
-		assert.Equal(t, test.expected, esScriptEnvVars(test.opts))
+		assert.EqualValues(t, test.expected, EsScriptEnvVars(test.opts))
 	}
 }
 

--- a/pkg/cronjob/spark_dependencies.go
+++ b/pkg/cronjob/spark_dependencies.go
@@ -12,7 +12,6 @@ import (
 
 	"github.com/jaegertracing/jaeger-operator/pkg/account"
 	v1 "github.com/jaegertracing/jaeger-operator/pkg/apis/jaegertracing/v1"
-	"github.com/jaegertracing/jaeger-operator/pkg/storage"
 	"github.com/jaegertracing/jaeger-operator/pkg/util"
 )
 
@@ -25,6 +24,7 @@ func SupportedStorage(storage string) bool {
 
 // CreateSparkDependencies creates a new cronjob for the Spark Dependencies task
 func CreateSparkDependencies(jaeger *v1.Jaeger) *batchv1beta1.CronJob {
+	logTLSNotSupported(jaeger)
 	envVars := []corev1.EnvVar{
 		{Name: "STORAGE", Value: jaeger.Spec.Storage.Type},
 		{Name: "SPARK_MASTER", Value: jaeger.Spec.Storage.Dependencies.SparkMaster},
@@ -32,16 +32,7 @@ func CreateSparkDependencies(jaeger *v1.Jaeger) *batchv1beta1.CronJob {
 	}
 	envVars = append(envVars, getStorageEnvs(jaeger.Spec.Storage)...)
 
-	var envFromSource []corev1.EnvFromSource
-	if len(jaeger.Spec.Storage.SecretName) > 0 {
-		envFromSource = append(envFromSource, corev1.EnvFromSource{
-			SecretRef: &corev1.SecretEnvSource{
-				LocalObjectReference: corev1.LocalObjectReference{
-					Name: jaeger.Spec.Storage.SecretName,
-				},
-			},
-		})
-	}
+	envFromSource := util.CreateEnvsFromSecret(jaeger.Spec.Storage.SecretName)
 
 	trueVar := true
 	one := int32(1)
@@ -94,7 +85,7 @@ func CreateSparkDependencies(jaeger *v1.Jaeger) *batchv1beta1.CronJob {
 									Image: jaeger.Spec.Storage.Dependencies.Image,
 									Name:  name,
 									// let spark job use its default values
-									Env:       removeEmptyVars(envVars),
+									Env:       util.RemoveEmptyVars(envVars),
 									EnvFrom:   envFromSource,
 									Resources: commonSpec.Resources,
 								},
@@ -117,14 +108,13 @@ func CreateSparkDependencies(jaeger *v1.Jaeger) *batchv1beta1.CronJob {
 }
 
 func getStorageEnvs(s v1.JaegerStorageSpec) []corev1.EnvVar {
-	sFlags := s.Options.Filter(storage.OptionsPrefix(s.Type))
-	sFlagsMap := sFlags.Map()
-	keyspace := sFlagsMap["cassandra.keyspace"]
-	if keyspace == "" {
-		keyspace = "jaeger_v1_test"
-	}
+	sFlagsMap := s.Options.Map()
 	switch s.Type {
 	case "cassandra":
+		keyspace := sFlagsMap["cassandra.keyspace"]
+		if keyspace == "" {
+			keyspace = "jaeger_v1_test"
+		}
 		return []corev1.EnvVar{
 			{Name: "CASSANDRA_CONTACT_POINTS", Value: sFlagsMap["cassandra.servers"]},
 			{Name: "CASSANDRA_KEYSPACE", Value: keyspace},
@@ -153,12 +143,12 @@ func getStorageEnvs(s v1.JaegerStorageSpec) []corev1.EnvVar {
 	}
 }
 
-func removeEmptyVars(envVars []corev1.EnvVar) []corev1.EnvVar {
-	var notEmpty []corev1.EnvVar
-	for _, v := range envVars {
-		if v.Value != "" || v.ValueFrom != nil {
-			notEmpty = append(notEmpty, v)
-		}
+func logTLSNotSupported(j *v1.Jaeger) {
+	sFlagsMap := j.Spec.Storage.Options.Map()
+	if strings.EqualFold(sFlagsMap["es.tls.enabled"], "true") || strings.EqualFold(sFlagsMap["es.tls"], "true") {
+		j.Logger().Warn("Spark dependencies does not support TLS with Elasticsearch, consider disabling dependencies")
 	}
-	return notEmpty
+	if strings.EqualFold(sFlagsMap["es.tls.skip-host-verify"], "true") || sFlagsMap["es.tls.ca"] != "" {
+		j.Logger().Warn("Spark dependencies does not support insecure TLS nor specifying only CA cert, consider disabling dependencies")
+	}
 }

--- a/pkg/cronjob/spark_dependencies_test.go
+++ b/pkg/cronjob/spark_dependencies_test.go
@@ -11,22 +11,6 @@ import (
 	v1 "github.com/jaegertracing/jaeger-operator/pkg/apis/jaegertracing/v1"
 )
 
-func TestRemoveEmptyVars(t *testing.T) {
-	tests := []struct {
-		underTest []corev1.EnvVar
-		expected  []corev1.EnvVar
-	}{
-		{},
-		{underTest: []corev1.EnvVar{{Name: "foo", Value: "bar"}, {Name: "foo3"}, {Name: "foo2", ValueFrom: &corev1.EnvVarSource{}}},
-			expected: []corev1.EnvVar{{Name: "foo", Value: "bar"}, {Name: "foo2", ValueFrom: &corev1.EnvVarSource{}}}},
-		{underTest: []corev1.EnvVar{{Name: "foo"}}},
-	}
-	for _, test := range tests {
-		exp := removeEmptyVars(test.underTest)
-		assert.Equal(t, test.expected, exp)
-	}
-}
-
 func TestStorageEnvs(t *testing.T) {
 	trueVar := true
 	falseVar := false

--- a/pkg/storage/elasticsearch_dependencies.go
+++ b/pkg/storage/elasticsearch_dependencies.go
@@ -9,6 +9,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	v1 "github.com/jaegertracing/jaeger-operator/pkg/apis/jaegertracing/v1"
+	"github.com/jaegertracing/jaeger-operator/pkg/cronjob"
 	"github.com/jaegertracing/jaeger-operator/pkg/util"
 )
 
@@ -20,30 +21,41 @@ func EnableRollover(spec v1.JaegerStorageSpec) bool {
 
 func elasticsearchDependencies(jaeger *v1.Jaeger) []batchv1.Job {
 	name := fmt.Sprintf("%s-es-rollover-create-mapping", jaeger.Name)
+	envFromSource := util.CreateEnvsFromSecret(jaeger.Spec.Storage.SecretName)
+	commonSpec := &v1.JaegerCommonSpec{
+		Annotations: map[string]string{
+			"prometheus.io/scrape":    "false",
+			"sidecar.istio.io/inject": "false",
+			"linkerd.io/inject":       "disabled",
+		},
+		Labels: util.Labels(name, "job-es-rollover-create-mapping", *jaeger),
+	}
+	commonSpec = util.Merge([]v1.JaegerCommonSpec{jaeger.Spec.Storage.EsRollover.JaegerCommonSpec, jaeger.Spec.JaegerCommonSpec, *commonSpec})
 	job := batchv1.Job{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:            name,
 			Namespace:       jaeger.Namespace,
-			Labels:          util.Labels(name, "job-es-rollover-create-mapping", *jaeger),
+			Labels:          commonSpec.Labels,
 			OwnerReferences: []metav1.OwnerReference{util.AsOwner(jaeger)},
 		},
 		Spec: batchv1.JobSpec{
 			Template: corev1.PodTemplateSpec{
 				ObjectMeta: metav1.ObjectMeta{
-					Annotations: map[string]string{
-						"prometheus.io/scrape":    "false",
-						"sidecar.istio.io/inject": "false",
-						"linkerd.io/inject":       "disabled",
-					},
+					Annotations: commonSpec.Annotations,
+					Labels:      commonSpec.Labels,
 				},
 				Spec: corev1.PodSpec{
 					RestartPolicy: corev1.RestartPolicyOnFailure,
+					Volumes:       commonSpec.Volumes,
 					Containers: []corev1.Container{
 						{
-							Name:  name,
-							Image: jaeger.Spec.Storage.EsRollover.Image,
-							Args:  []string{"init", util.GetEsHostname(jaeger.Spec.Storage.Options.Map())},
-							Env:   envVars(jaeger.Spec.Storage.Options),
+							Name:         name,
+							Image:        jaeger.Spec.Storage.EsRollover.Image,
+							Args:         []string{"init", util.GetEsHostname(jaeger.Spec.Storage.Options.Map())},
+							Env:          util.RemoveEmptyVars(envVars(jaeger.Spec.Storage.Options)),
+							EnvFrom:      envFromSource,
+							Resources:    commonSpec.Resources,
+							VolumeMounts: commonSpec.VolumeMounts,
 						},
 					},
 				},
@@ -53,22 +65,20 @@ func elasticsearchDependencies(jaeger *v1.Jaeger) []batchv1.Job {
 	return []batchv1.Job{job}
 }
 
-func envVars(spec v1.Options) []corev1.EnvVar {
-	var envs []corev1.EnvVar
-	if val, ok := spec.Map()["es.index-prefix"]; ok {
-		envs = append(envs, corev1.EnvVar{Name: "INDEX_PREFIX", Value: val})
+func envVars(opts v1.Options) []corev1.EnvVar {
+	var envs = cronjob.EsScriptEnvVars(opts)
+	scriptEnvVars := []struct {
+		flag   string
+		envVar string
+	}{
+		{flag: "es.num-shards", envVar: "SHARDS"},
+		{flag: "es.num-replicas", envVar: "REPLICAS"},
 	}
-	if val, ok := spec.Map()["es.num-shards"]; ok {
-		envs = append(envs, corev1.EnvVar{Name: "SHARDS", Value: val})
-	}
-	if val, ok := spec.Map()["es.num-replicas"]; ok {
-		envs = append(envs, corev1.EnvVar{Name: "REPLICAS", Value: val})
-	}
-	if val, ok := spec.Map()["es.username"]; ok {
-		envs = append(envs, corev1.EnvVar{Name: "ES_USERNAME", Value: val})
-	}
-	if val, ok := spec.Map()["es.password"]; ok {
-		envs = append(envs, corev1.EnvVar{Name: "ES_PASSWORD", Value: val})
+	options := opts.Map()
+	for _, x := range scriptEnvVars {
+		if val, ok := options[x.flag]; ok {
+			envs = append(envs, corev1.EnvVar{Name: x.envVar, Value: val})
+		}
 	}
 	return envs
 }

--- a/pkg/storage/elasticsearch_dependencies_test.go
+++ b/pkg/storage/elasticsearch_dependencies_test.go
@@ -86,10 +86,11 @@ func TestEnvVars(t *testing.T) {
 				"es.username":     "fredy"}),
 			expected: []corev1.EnvVar{
 				{Name: "INDEX_PREFIX", Value: "foo"},
+				{Name: "ES_USERNAME", Value: "fredy"},
+				{Name: "ES_PASSWORD", Value: "nopass"},
 				{Name: "SHARDS", Value: "5"},
 				{Name: "REPLICAS", Value: "3"},
-				{Name: "ES_USERNAME", Value: "fredy"},
-				{Name: "ES_PASSWORD", Value: "nopass"}},
+			},
 		},
 	}
 	for _, test := range tests {

--- a/pkg/strategy/controller.go
+++ b/pkg/strategy/controller.go
@@ -107,10 +107,18 @@ func distributedStorage(storage string) bool {
 }
 
 func normalizeSparkDependencies(spec *v1.JaegerStorageSpec) {
+	sFlagsMap := spec.Options.Map()
+	tlsEnabled := sFlagsMap["es.tls"]
+	tlsSkipHost := sFlagsMap["es.tls.skip-host-verify"]
+	tlsCa := sFlagsMap["es.tls.ca"]
+	tlsIsNotEnabled := !strings.EqualFold(tlsEnabled, "true") &&
+		!strings.EqualFold(tlsSkipHost, "true") &&
+		strings.EqualFold(tlsCa, "")
 	// auto enable only for supported storages
 	if cronjob.SupportedStorage(spec.Type) &&
 		spec.Dependencies.Enabled == nil &&
-		!storage.ShouldDeployElasticsearch(*spec) {
+		!storage.ShouldDeployElasticsearch(*spec) &&
+		tlsIsNotEnabled {
 		trueVar := true
 		spec.Dependencies.Enabled = &trueVar
 	}

--- a/pkg/strategy/controller_test.go
+++ b/pkg/strategy/controller_test.go
@@ -237,6 +237,27 @@ func TestNormalizeSparkDependencies(t *testing.T) {
 			expected:  v1.JaegerStorageSpec{Type: "elasticsearch", Dependencies: v1.JaegerDependenciesSpec{Schedule: "55 23 * * *", Image: "foo"}},
 		},
 		{
+			underTest: v1.JaegerStorageSpec{Type: "elasticsearch", Dependencies: v1.JaegerDependenciesSpec{},
+				Options: v1.NewOptions(map[string]interface{}{"es.server-urls": "local", "es.tls": true})},
+			expected: v1.JaegerStorageSpec{Type: "elasticsearch", Dependencies: v1.JaegerDependenciesSpec{Image: "foo", Schedule: "55 23 * * *", Enabled: nil},
+				Options: v1.NewOptions(map[string]interface{}{"es.server-urls": "local", "es.tls": true}),
+			},
+		},
+		{
+			underTest: v1.JaegerStorageSpec{Type: "elasticsearch", Dependencies: v1.JaegerDependenciesSpec{},
+				Options: v1.NewOptions(map[string]interface{}{"es.server-urls": "local", "es.skip-host-verify": false})},
+			expected: v1.JaegerStorageSpec{Type: "elasticsearch", Dependencies: v1.JaegerDependenciesSpec{Image: "foo", Schedule: "55 23 * * *", Enabled: &trueVar},
+				Options: v1.NewOptions(map[string]interface{}{"es.server-urls": "local", "es.skip-host-verify": false}),
+			},
+		},
+		{
+			underTest: v1.JaegerStorageSpec{Type: "elasticsearch", Dependencies: v1.JaegerDependenciesSpec{},
+				Options: v1.NewOptions(map[string]interface{}{"es.server-urls": "local", "es.skip-host-verify": false, "es.tls.ca": "rr"})},
+			expected: v1.JaegerStorageSpec{Type: "elasticsearch", Dependencies: v1.JaegerDependenciesSpec{Image: "foo", Schedule: "55 23 * * *", Enabled: nil},
+				Options: v1.NewOptions(map[string]interface{}{"es.server-urls": "local", "es.skip-host-verify": false, "es.tls.ca": "rr"}),
+			},
+		},
+		{
 			underTest: v1.JaegerStorageSpec{Type: "elasticsearch", Dependencies: v1.JaegerDependenciesSpec{Schedule: "foo", Image: "bla", Enabled: &falseVar}},
 			expected:  v1.JaegerStorageSpec{Type: "elasticsearch", Dependencies: v1.JaegerDependenciesSpec{Schedule: "foo", Image: "bla", Enabled: &falseVar}},
 		},

--- a/pkg/util/util.go
+++ b/pkg/util/util.go
@@ -216,3 +216,29 @@ func ImageName(image, param string) string {
 	}
 	return image
 }
+
+// RemoveEmptyVars removes empty variables from the input slice.
+func RemoveEmptyVars(envVars []corev1.EnvVar) []corev1.EnvVar {
+	var notEmpty []corev1.EnvVar
+	for _, v := range envVars {
+		if v.Value != "" || v.ValueFrom != nil {
+			notEmpty = append(notEmpty, v)
+		}
+	}
+	return notEmpty
+}
+
+// CreateEnvsFromSecret adds env from secret name.
+func CreateEnvsFromSecret(secretName string) []corev1.EnvFromSource {
+	var envs []corev1.EnvFromSource
+	if len(secretName) > 0 {
+		envs = append(envs, corev1.EnvFromSource{
+			SecretRef: &corev1.SecretEnvSource{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: secretName,
+				},
+			},
+		})
+	}
+	return envs
+}

--- a/pkg/util/util_test.go
+++ b/pkg/util/util_test.go
@@ -424,3 +424,37 @@ func TestImageNameParamDefaultWithTag(t *testing.T) {
 
 	assert.Equal(t, "org/default-image:1.2.3", ImageName("", "test-image"))
 }
+
+func TestRemoveEmptyVars(t *testing.T) {
+	tests := []struct {
+		underTest []corev1.EnvVar
+		expected  []corev1.EnvVar
+	}{
+		{},
+		{underTest: []corev1.EnvVar{{Name: "foo", Value: "bar"}, {Name: "foo3"}, {Name: "foo2", ValueFrom: &corev1.EnvVarSource{}}},
+			expected: []corev1.EnvVar{{Name: "foo", Value: "bar"}, {Name: "foo2", ValueFrom: &corev1.EnvVarSource{}}}},
+		{underTest: []corev1.EnvVar{{Name: "foo"}}},
+	}
+	for _, test := range tests {
+		exp := RemoveEmptyVars(test.underTest)
+		assert.Equal(t, test.expected, exp)
+	}
+}
+
+func TestCreateFromSecret(t *testing.T) {
+	tests := []struct {
+		secret   string
+		expected []corev1.EnvFromSource
+	}{
+		{},
+		{
+			secret: "foobar", expected: []corev1.EnvFromSource{
+				{SecretRef: &corev1.SecretEnvSource{LocalObjectReference: corev1.LocalObjectReference{Name: "foobar"}}},
+			},
+		},
+	}
+	for _, test := range tests {
+		exp := CreateEnvsFromSecret(test.secret)
+		assert.Equal(t, test.expected, exp)
+	}
+}

